src/Workspaces/Core/Portable/EmbeddedLanguages/DateTime/LanguageServices/DateTimePatternDetector.cs

```diff
@@ -0,0 +1,202 @@
+            if (dateTimeType == null || dateTimeOffsetType == null)
+                return null;
+
+            return new DateTimePatternDetector(semanticModel, info, dateTimeType, dateTimeOffsetType);
+        }
+
+        public static bool IsDefinitelyNotDateTimeStringToken(SyntaxToken token, ISyntaxFacts syntaxFacts)
+        {
+            if (!syntaxFacts.IsStringLiteral(token))
+                return true;
+
+            if (!IsMethodArgument(token, syntaxFacts))
+                return true;
+
+            return false;
+        }
+
+        private static bool IsMethodArgument(SyntaxToken token, ISyntaxFacts syntaxFacts)
+            => syntaxFacts.IsLiteralExpression(token.Parent) &&
+               syntaxFacts.IsArgument(token.Parent!.Parent);
+
+        public bool IsDateTimeStringToken(SyntaxToken token, CancellationToken cancellationToken)
+        {
+            if (IsDefinitelyNotDateTimeStringToken(token, _info.SyntaxFacts))
+                return false;
+
+            var syntaxFacts = _info.SyntaxFacts;
+
+            var stringLiteral = token;
+            var literalNode = stringLiteral.Parent;
+            var argumentNode = literalNode!.Parent;
+            Debug.Assert(syntaxFacts.IsArgument(argumentNode));
+
+            var argumentList = argumentNode!.Parent;
+            var invocationOrCreation = argumentList!.Parent;
+            if (!syntaxFacts.IsInvocationExpression(invocationOrCreation))
+                return false;
+
+            var invokedExpression = syntaxFacts.GetExpressionOfInvocationExpression(invocationOrCreation);
+            var name = GetNameOfInvokedExpression(invokedExpression);
+            if (name != nameof(ToString) && name != nameof(System.DateTime.ParseExact) && name != nameof(System.DateTime.TryParseExact))
+                return false;
+
+            // if we couldn't determine the arg name or arg index, can't proceed.
+            var (argName, argIndex) = GetArgumentNameOrIndex(argumentNode);
+            if (argName == null && argIndex == null)
+                return false;
+
+            // If we had a specified arg name and it isn't 'format', then it's not a DateTime
+            // 'format' param we care about.
+            if (argName != null && argName != FormatName)
+                return false;
+
+            var symbolInfo = _semanticModel.GetSymbolInfo(invocationOrCreation, cancellationToken);
+            var method = symbolInfo.Symbol;
+            if (TryAnalyzeInvocation(method, argName, argIndex))
+                return true;
+
+            foreach (var candidate in symbolInfo.CandidateSymbols)
+            {
+                if (TryAnalyzeInvocation(candidate, argName, argIndex))
+                    return true;
+            }
+
+            return false;
+        }
+
+        private (string? name, int? index) GetArgumentNameOrIndex(SyntaxNode argument)
```

Comment:
I think this would make much more sense to me if this was split into two methods, one for argument name, one for index.  I guess I should trust the compiler but it was a little confusing reading `AnalyzeStringLiteral` and trying to track down which one is null, which one is not